File: packages/common/models/slide.py
"""
Slide model - represents a single slide in a presentation
"""

from typing import TYPE_CHECKING, Any, Literal

from sqlalchemy import JSON, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from packages.common.models.base import Base, TimestampMixin

if TYPE_CHECKING:
    from packages.common.models.presentation import Presentation


SlideType = Literal["title", "content", "bullets", "quote", "section", "chart"]
SlideLayout = Literal["left", "center", "right", "split"]
ChartType = Literal["bar", "line", "pie", "donut", "area", "horizontal_bar"]


class Slide(Base, TimestampMixin):
    """
    Slide model representing a single slide in a presentation.

    Attributes:
        id: Primary key
        presentation_id: Foreign key to parent presentation
        type: Slide type (title, content, bullets, quote, section, chart)
        title: Slide title
        subtitle: Optional subtitle
        body: Main text content
        bullets: List of bullet points (JSON)
        quote: Quote text for quote slides
        attribution: Quote attribution
        layout: Layout type (left, center, right, split)
        order: Position in presentation
        chart_type: Type of chart (bar, line, pie, donut, area, horizontal_bar)
        chart_data: Chart data points (JSON array)
        chart_config: Chart configuration options (JSON object)
    """

    __tablename__ = "slides"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    presentation_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("presentations.id", ondelete="CASCADE"),
        nullable=False,
    )

    # Slide content
    type: Mapped[str] = mapped_column(String(50), default="content", nullable=False)
    title: Mapped[str | None] = mapped_column(String(255), nullable=True)
    subtitle: Mapped[str | None] = mapped_column(String(255), nullable=True)
    body: Mapped[str | None] = mapped_column(Text, nullable=True)
    bullets: Mapped[list[str] | None] = mapped_column(JSON, nullable=True)
    quote: Mapped[str | None] = mapped_column(Text, nullable=True)
    attribution: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # Chart fields
    chart_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    chart_data: Mapped[list[dict[str, Any]] | None] = mapped_column(JSON, nullable=True)
    chart_config: Mapped[dict[str, Any] | None] = mapped_column(JSON, nullable=True)

    # Layout
    layout: Mapped[str] = mapped_column(String(50), default="center", nullable=False)
    order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)

    # Relationships
    presentation: Mapped["Presentation"] = relationship(
        "Presentation",
        back_populates="slides",
    )

    def __repr__(self) -> str:
        return f"<Slide(id={self.id}, type='{self.type}', order={self.order})>"

File: apps/frontend/lib/types/slide.ts
/**
 * Slide and Presentation types
 * Mirrors backend Pydantic schemas
 */

export type SlideType = "title" | "content" | "bullets" | "quote" | "section" | "chart";
export type SlideLayout = "left" | "center" | "right" | "split";
export type ChartType = "bar" | "line" | "pie" | "donut" | "area" | "horizontal_bar";
export type ThemeName = "neobrutalism" | "corporate" | "minimal" | "dark" | "magazine" | "terminal" | "playful";

// =============================================================================
// THEME TYPE DEFINITIONS
// =============================================================================

export interface ThemeColors {
  background: string;
  surface: string;
  text_primary: string;
  text_secondary: string;
  accent: string;
  accent_hover: string;
  accent_light: string;
  border: string;
  border_dark: string;
}

export interface ThemeTypography {
  heading_font: string;
  body_font: string;
  google_fonts: string[];
  title_size: string;
  title_weight: number;
  title_letter_spacing: string;
  title_transform: "none" | "uppercase" | "lowercase" | "capitalize";
  title_line_height: number;
  heading_size: string;
  heading_weight: number;
  heading_letter_spacing: string;
  body_size: string;
  body_weight: number;
  body_line_height: number;
  quote_size: string;
  quote_style: "normal" | "italic";
}

export interface ThemeStyle {
  border_width: string;
  border_style: "solid" | "dashed" | "dotted" | "none";
  border_radius: string;
  shadow: string | null;
  accent_bar_position: "none" | "top" | "left" | "bottom";
  accent_bar_width: string;
  background_pattern: "none" | "dots" | "grid" | "scanlines" | "noise";
  pattern_opacity: number;
}

export interface ThemeSpacing {
  slide_padding: string;
  element_gap: string;
  bullet_gap: string;
  content_max_width: string;
}

export interface ThemeDecorations {
  bullet_style: "disc" | "dash" | "arrow" | "number" | "check" | "square";
  bullet_size: string;
  quote_style: "large-mark" | "accent-bar" | "icon" | "none";
  section_divider: "line" | "accent-block" | "gradient" | "none";
}

export interface ThemeLayout {
  title_alignment: "left" | "center" | "right";
  content_alignment: "left" | "center" | "right";
  vertical_position: "top" | "center" | "bottom";
}

export interface ThemeDefinition {
  name: ThemeName;
  display_name: string;
  description: string;
  colors: ThemeColors;
  typography: ThemeTypography;
  style: ThemeStyle;
  spacing: ThemeSpacing;
  decorations: ThemeDecorations;
  layout: ThemeLayout;
}

export interface ChartDataPoint {
  label: string;
  value: number;
  color?: string;
}

export interface ChartConfig {
  show_legend?: boolean;
  show_values?: boolean;
  y_axis_label?: string;
  x_axis_label?: string;
}

export interface Slide {
  id: number;
  type: SlideType;
  title?: string | null;
  subtitle?: string | null;
  body?: string | null;
  bullets?: string[] | null;
  quote?: string | null;
  attribution?: string | null;
  layout: SlideLayout;
  order: number;
  // Chart fields
  chart_type?: ChartType | null;
  chart_data?: ChartDataPoint[] | null;
  chart_config?: ChartConfig | null;
}

export interface Presentation {
  id: number;
  title: string;
  input_text: string;
  theme: string;
  slides: Slide[];
  created_at: string;
  updated_at: string;
}

export interface GenerateSlidesRequest {
  text: string;
  slide_count?: number;
  title?: string;
  theme?: ThemeName;
}

export interface GenerateSlidesResponse {
  presentation: Presentation;
}

export interface SlideUpdate {
  type?: SlideType;
  title?: string | null;
  subtitle?: string | null;
  body?: string | null;
  bullets?: string[] | null;
  quote?: string | null;
  attribution?: string | null;
  layout?: SlideLayout;
  // Chart fields
  chart_type?: ChartType | null;
  chart_data?: ChartDataPoint[] | null;
  chart_config?: ChartConfig | null;
}

export interface AgentEvent {
  type: "thinking" | "tool_call" | "tool_result" | "complete" | "error";
  message?: string;
  tool?: string;
  args?: Record<string, unknown>;
  slide_number?: number;
  slide?: Slide;
  result?: string;
  success?: boolean;
  presentation_id?: number;
  title?: string;
  slide_count?: number;
  iteration?: number;
  presentation?: Presentation;
}
